components/board/genericlinux/hw_pwm.go

```diff
@@ -0,0 +1,174 @@
+	// There should be a single directory within /sys/devices/platform/<chipName>/pwm/, whose name
+	// is mirrored in /sys/class/pwm. That's the one we want to use.
+	chipDir := fmt.Sprintf("/sys/devices/platform/%s/pwm", chipName)
+	files, err := os.ReadDir(chipDir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, file := range files {
+		if strings.Contains(file.Name(), "pwmchip") && file.IsDir() {
+			chipPath := fmt.Sprintf("/sys/class/pwm/%s", file.Name())
+			return &pwmDevice{chipPath: chipPath, line: line}, nil
+		}
+	}
+	return nil, errors.Errorf("Could not find any PWM device with name %s", chipName)
+}
+
+func writeValue(filepath string, value uint64) error {
+	// The permissions on the file (the third argument) aren't important: if the file needs to be
+	// created, something has gone horribly wrong!
+	return os.WriteFile(filepath, []byte(fmt.Sprintf("%d", value)), 0o600)
+}
+
+func (pwm *pwmDevice) writeChip(filename string, value uint64) error {
+	return writeValue(fmt.Sprintf("%s/%s", pwm.chipPath, filename), value)
+}
+
+func (pwm *pwmDevice) writeLine(filename string, value uint64) error {
+	return writeValue(fmt.Sprintf("%s/pwm%d/%s", pwm.chipPath, pwm.line, filename), value)
+}
+
+// Export tells the OS that this pin is in use, and enables configuration via sysfs.
+func (pwm *pwmDevice) export() error {
+	if pwm.isExported {
+		return nil // Already exported
+	}
+	if err := pwm.writeChip("export", uint64(pwm.line)); err != nil {
+		return err
+	}
+	pwm.isExported = true
+	return nil
+}
+
+// Unexport turns off any PWM signal the pin was providing, and tells the OS that this pin is no
+// longer in use (so it can be reused as an input pin, etc.).
+func (pwm *pwmDevice) unexport() error {
+	if !pwm.isExported {
+		return nil // Already unexported
+	}
+	// If we unexport the pin while it is enabled, it might continue outputting a PWM signal,
+	// causing trouble if you start using the pin for something else.
+	if err := pwm.disable(); err != nil {
+		return err
+	}
+	if err := pwm.writeChip("unexport", uint64(pwm.line)); err != nil {
+		return err
+	}
+	pwm.isExported = false
+	return nil
+}
+
+// Enable tells an exported pin to output the PWM signal it has been configured with.
+func (pwm *pwmDevice) enable() error {
+	if pwm.isEnabled {
+		return nil // Already enabled
+	}
+	if err := pwm.writeLine("enable", 1); err != nil {
+		return err
+	}
+	pwm.isEnabled = true
+	return nil
+}
+
+// Disable tells an exported pin to stop outputting its PWM signal, but it is still available for
+// reconfiguring and re-enabling.
+func (pwm *pwmDevice) disable() error {
+	if !pwm.isEnabled {
+		return nil // Already disabled
+	}
+	if err := pwm.writeLine("enable", 0); err != nil {
+		return err
+	}
+	pwm.isEnabled = false
+	return nil
+}
+
+// SetPwm configures an exported pin and enables its output signal.
+// Warning: if this function returns a non-nil error, it could leave the pin in an indeterminate
```

Comment:
[q] is there a way we can return the pin to a determinate state if an error has been detected? For example, if it has succeeded in writing the duty cycle but failed to write the period, perhaps it should write the duty cycle back to the old value before raising the error

Reply:
What's the use case for this? What kind of errors might we encounter where we can automatically set things right again afterwards? The errors I can think of are "you don't have permission to interact with hardware; rerun the server as `root`," and "the pin is already in use by a different device; go disable the I2C bus." Neither of those would be helped by recovering to a known state. 

Without a use case in mind, I worry that doing this could be a lot of effort for no gain.

Reply:
upon reflection, I agree. This comment can be ignored